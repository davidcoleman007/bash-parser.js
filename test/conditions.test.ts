import { describe, expect, it } from 'vitest';
import bashParser from '~/parse.ts';

describe('conditions', async (t) => {
  it('if [ "$a" -eq 10 ]', async (t) => {
    expect(await bashParser('if [ "$a" -eq 10 ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ "$a" -ne 10 ]', async (t) => {
    expect(await bashParser('if [ "$a" -ne 10 ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('[ "$a" -gt 10 ]', async (t) => {
    expect(await bashParser('if [ "$a" -gt 10 ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ "$a" -ge 10 ]', async (t) => {
    expect(await bashParser('if [ "$a" -ge 10 ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('[ "$a" -lt 10 ]', async (t) => {
    expect(await bashParser('if [ "$a" -lt 10 ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ "$a" -le 10 ]', async (t) => {
    expect(await bashParser('if [ "$a" -le 10 ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ "$a" == 10 ]', async (t) => {
    expect(await bashParser('if [ "$a" == 10 ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ "$a" != 10 ]', async (t) => {
    expect(await bashParser('if [ "$a" != 10 ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ ! "$a" = 10 ]', async (t) => {
    expect(await bashParser('if [ ! "$a" = 10 ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ -z "$a" ]', async (t) => {
    expect(await bashParser('if [ -z "$a" ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ -n "$a" ]', async (t) => {
    expect(await bashParser('if [ -n "$a" ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ -d "$a" ]', async (t) => {
    expect(await bashParser('if [ -d "$a" ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ -f "$a" ]', async (t) => {
    expect(await bashParser('if [ -f "$a" ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ -r "$a" ]', async (t) => {
    expect(await bashParser('if [ -r "$a" ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ -w "$a" ]', async (t) => {
    expect(await bashParser('if [ -w "$a" ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('if [ -x "$a" ]', async (t) => {
    expect(await bashParser('if [ -x "$a" ]; then echo 1; fi')).toMatchSnapshot();
  });

  it('while [ "$a" -le 10 ]', async (t) => {
    expect(await bashParser('while [ "$a" -le 10 ]; do echo 1; done')).toMatchSnapshot();
  });

  it('until [ "$a" -ge 10 ]', async (t) => {
    expect(await bashParser('until [ "$a" -ge 10 ]; do echo 1; done')).toMatchSnapshot();
  });

  it('if [ "$(command)" -eq 1 ]', async (t) => {
    expect(await bashParser('if [ "$(command)" -eq 1 ]; then echo 1; fi')).toMatchSnapshot();
  });
});
