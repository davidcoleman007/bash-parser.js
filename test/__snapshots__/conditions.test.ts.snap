export const snapshot = {};

snapshot[`conditions > if [ "\$a" -eq 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "-eq",
          type: "Equal",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ "\$a" -ne 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "-ne",
          type: "NotEqual",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > [ "\$a" -gt 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "-gt",
          type: "GreaterThan",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ "\$a" -ge 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "-ge",
          type: "GreaterThanOrEqual",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > [ "\$a" -lt 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "-lt",
          type: "LessThan",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ "\$a" -le 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "-le",
          type: "LessThanOrEqual",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ "\$a" == 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "==",
          type: "Equal",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ "\$a" != 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "!=",
          type: "NotEqual",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ ! "\$a" = 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        inverted: true,
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "=",
          type: "Equal",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ -z "\$a" ] 1`] = `
{
  commands: [
    {
      clause: {
        op: {
          text: "-z",
          type: "IsStringNull",
        },
        target: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        type: "TestExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ -n "\$a" ] 1`] = `
{
  commands: [
    {
      clause: {
        op: {
          text: "-n",
          type: "IsStringNotNull",
        },
        target: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        type: "TestExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ -d "\$a" ] 1`] = `
{
  commands: [
    {
      clause: {
        op: {
          text: "-d",
          type: "IsDirectory",
        },
        target: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        type: "TestExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ -f "\$a" ] 1`] = `
{
  commands: [
    {
      clause: {
        op: {
          text: "-f",
          type: "IsFile",
        },
        target: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        type: "TestExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ -r "\$a" ] 1`] = `
{
  commands: [
    {
      clause: {
        op: {
          text: "-r",
          type: "IsReadable",
        },
        target: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        type: "TestExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ -w "\$a" ] 1`] = `
{
  commands: [
    {
      clause: {
        op: {
          text: "-w",
          type: "IsWritable",
        },
        target: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        type: "TestExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ -x "\$a" ] 1`] = `
{
  commands: [
    {
      clause: {
        op: {
          text: "-x",
          type: "IsExecutable",
        },
        target: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        type: "TestExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > while [ "\$a" -le 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "-le",
          type: "LessThanOrEqual",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      do: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "While",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > until [ "\$a" -ge 10 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              loc: {
                end: 2,
                start: 1,
              },
              parameter: "a",
              type: "ParameterExpansion",
            },
          ],
          text: '"\$a"',
          type: "Word",
        },
        op: {
          text: "-ge",
          type: "GreaterThanOrEqual",
        },
        right: {
          text: "10",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      do: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "Until",
    },
  ],
  type: "Script",
}
`;

snapshot[`conditions > if [ "\$(command)" -eq 1 ] 1`] = `
{
  commands: [
    {
      clause: {
        left: {
          expansion: [
            {
              command: "command",
              commandAST: {
                commands: [
                  {
                    name: {
                      text: "command",
                      type: "Word",
                    },
                    type: "Command",
                  },
                ],
                type: "Script",
              },
              loc: {
                end: 10,
                start: 1,
              },
              type: "CommandExpansion",
            },
          ],
          text: '"\$(command)"',
          type: "Word",
        },
        op: {
          text: "-eq",
          type: "Equal",
        },
        right: {
          text: "1",
          type: "Word",
        },
        type: "LogicalExpression",
      },
      then: {
        commands: [
          {
            name: {
              text: "echo",
              type: "Word",
            },
            suffix: [
              {
                text: "1",
                type: "Word",
              },
            ],
            type: "Command",
          },
        ],
        type: "CompoundList",
      },
      type: "If",
    },
  ],
  type: "Script",
}
`;
